### DISCUSSION TITLE
Best practice to increase build performance for large solutions

### DISCUSSION THREAD

**Discussion by Sam13** [dotnet/roslyn#45933] [General] [answered]
I have a large VS 2019 (16.6.3) solution which contains around 160 project files (C#, .NET 4.8, SDK style project format).
I'm using StyleCop, FxCop and custom analyzers which are referenced as NuGet package in the _Directory.build.props_ file.

Build time significantly increases when having e.g. _Microsoft.CodeAnalysis.FxCopAnalyzers_ enabled.

This leads to the following question:
Is there any recommended practice to increase build performance during development work in Visual Studio and still have the opportunity to run the analyzers manually?

I tried to add the following line into _Directory.build.props_
```
<RunAnalyzersDuringBuild Condition="'$(Configuration)' == 'Debug' AND '$(BuildingInsideVisualStudio)' == 'true'">False</RunAnalyzersDuringBuild>
```
When having that, manually running code analysis over via menu bar (_Analyze -> Run code analysis -> On Solution_) does no longer work...

**Comment by jmarolf** [comment]
CC: @mavasani

**Comment by mavasani** [comment]
> Build time significantly increases when having e.g. Microsoft.CodeAnalysis.FxCopAnalyzers enabled.

Can you try using the `3.3.0-beta1.final` version https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers/3.3.0-beta1.final, which has many performance fixes? Also, would you be able to submit a performance trace for your performance issue by following the steps at https://aka.ms/reportPerf?

> Is there any recommended practice to increase build performance during development work in Visual Studio and still have the opportunity to run the analyzers manually?

Can you try to turn off `RunAnalyzersDuringLiveAnalysis` instead? After that you should be able to manually trigger on-deman code analysis by executing `Run Code Analysis` command. See https://docs.microsoft.com/visualstudio/code-quality/disable-code-analysis?view=vs-2019#net-core-and-net-standard-projects and https://docs.microsoft.com/visualstudio/code-quality/how-to-run-code-analysis-manually-for-managed-code?view=vs-2019.
Another thing you can try is to switch your background analysis scope to `Current Document` only: https://docs.microsoft.com/visualstudio/code-quality/configure-live-code-analysis-scope-managed-code?view=vs-2019#custom-analysis-scope

**Reply by Sam13** [reply]
> > Build time significantly increases when having e.g. Microsoft.CodeAnalysis.FxCopAnalyzers enabled.
> 
> Can you try using the `3.3.0-beta1.final` version https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers/3.3.0-beta1.final, which has many performance fixes?

I did some quick tests: My 160+ projects solution builds in 1:42 minutes without any analyzers.
With StyleCop + FxCop v3.0.0 in 8:00 and with StyleCop + FxCop v3.3.0 beta 1 in 7:55 - not a big change there. May have to investigate into that further...

> Can you try to turn off `RunAnalyzersDuringLiveAnalysis` instead? After that you should be able to manually trigger on-deman code analysis by executing `Run Code Analysis` command. See https://docs.microsoft.com/visualstudio/code-quality/disable-code-analysis?view=vs-2019#net-core-and-net-standard-projects and https://docs.microsoft.com/visualstudio/code-quality/how-to-run-code-analysis-manually-for-managed-code?view=vs-2019.
> Another thing you can try is to switch your background analysis scope to `Current Document` only: https://docs.microsoft.com/visualstudio/code-quality/configure-live-code-analysis-scope-managed-code?view=vs-2019#custom-analysis-scope

How should that affect my build time? I'm happy with the overall VS performance but not with the build time. E.g. when I change something in the production code and waiting for the unit test project to compile and execute...

**Comment by mavasani** [comment]
would you be able to submit a performance trace for your performance issue by following the steps at https://aka.ms/reportPerf?

**Comment by Sam13** [comment, marked as answer]
@mavasani Thanks for your assistance. It's no longer necessary to provide a performance trace. After upgrading to FxCop analyzers v3.3.0-beta2.final the performance is fine again.
